lib/IRGen/IRGenModule.cpp

```diff
@@ -1185,37 +1175,134 @@ static bool isFirstObjectFileInModule(IRGenModule &IGM) {
   return containingModule->getFiles().front() == file;
 }
 
+static bool
+doesTargetAutolinkUsingAutolinkExtract(const SwiftTargetInfo &TargetInfo,
+                                       const llvm::Triple &Triple) {
+  if (TargetInfo.OutputObjectFormat == llvm::Triple::ELF && !Triple.isPS4())
+    return true;
+
+  if (TargetInfo.OutputObjectFormat == llvm::Triple::Wasm)
+    return true;
+
+  if (Triple.isOSCygMing())
+    return true;
+
+  return false;
+}
+
+namespace {
+enum class AutolinkKind {
```

Comment:
The nice way of doing this would be to use a struct enum pattern. The way you do this is that you:

```
struct AutolinkKind {
   enum ValueTy {
      ...
   };

   ValueTy value;

   AutolinkKind(ValueTy value) : Value(value) {}
   AutolinkKind(const AutolinkKind &kind) : Value(kind.value) {}
}
```

Then you can sink a lot of this logic into method names on Autolink. In fact you could even have private constructors and make determining the Kind a class method that selects the relevant value.